### ISSUE TITLE
Regression in multi-node training speed with Transformers + PyTorch

### ISSUE THREAD

**Issue by mvsjober** [pytorch/pytorch#71855]
### 🐛 Describe the bug

We have seen a regression in multi-node training speed with the [Huggingface Transformers CLM test case](https://github.com/huggingface/transformers/blob/master/examples/pytorch/language-modeling/run_clm.py) happening at PyTorch commit [38ac9e6](https://github.com/pytorch/pytorch/commit/38ac9e6) which reverts commit [3957ed4](https://github.com/pytorch/pytorch/commit/3957ed4) "[DDP] Disable reducer hooks from running outside of DDP backwards". Between those two commits I get around 28 `train_samples_per_second` (as reported by the CLM test) using two nodes with 4xNVIDIA A100 GPUs each. Afterwards speed drops to around 10 `train_samples_per_second`. This can be reliably reproduced every time.

My testing script can be found here, although it's a bit difficult to reproduce at it requires building the singularity containers and a cluster with at least two nodes: https://github.com/mvsjober/pytorch-clm-test-case/blob/main/pytorch_clm_test.sh

The PyTorch container has been created with [this build recipe](https://github.com/mvsjober/pytorch-clm-test-case/blob/main/pytorch_git_custom.def). It's a CentosOS 7.7 image with CUDA 11.1, cuDNN 8.1.0, NCCL 2.8.4.

### Versions

Collecting environment information...
PyTorch version: 1.10.0a0+git38ac9e6
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)
Clang version: Could not collect
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.8.11 (default, Sep  1 2021, 12:33:46)  [GCC 9.3.1 20200408 (Red Hat 9.3.1-2)] (64-bit runtime)
Python platform: Linux-3.10.0-1160.41.1.el7.x86_64-x86_64-with-glibc2.2.5
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.1.0
/usr/lib64/libcudnn_adv_infer.so.8.1.0
/usr/lib64/libcudnn_adv_train.so.8.1.0
/usr/lib64/libcudnn_cnn_infer.so.8.1.0
/usr/lib64/libcudnn_cnn_train.so.8.1.0
/usr/lib64/libcudnn_ops_infer.so.8.1.0
/usr/lib64/libcudnn_ops_train.so.8.1.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.22.1
[pip3] torch==1.10.0a0+git38ac9e6
[pip3] torchaudio==0.9.1
[pip3] torchtext==0.10.1
[pip3] torchvision==0.10.1+cu111
[conda] Could not collect

**Comment by rohan-varma** [comment]
Adding high priority due to perf regression

**Comment by zhaojuanmao** [comment]
@mvsjober is the regression still happening? also wondering what is the QPS before commit [3957ed4](https://github.com/pytorch/pytorch/commit/3957ed4)? 

If [3957ed4](https://github.com/pytorch/pytorch/commit/3957ed4) does help the QPS improvement, that mostly means this commit disabled the grad sync when directly using the local module wrapped by DDP, is that what you want in the application?

**Comment by mvsjober** [comment]
@zhaojuanmao Yes, regression is still happening. I just ran my test with the most recent release and I got around 8 `train_samples_per_second` and I also tried to most recent commit (which was c6f1bbc) and I got around 13. Compared to around 28 after commit 3957ed4.

And yes, before that "magic" commit it was also low. According to my notes with the previous commit 5a4282d it was 10.36.

I'm not sure regarding the grad sync, the application is the [Huggingface Transformers CLM test case](https://github.com/huggingface/transformers/blob/master/examples/pytorch/language-modeling/run_clm.py) so perhaps this should be brought up with them?
